### ISSUE TITLE
Track HTML element popularity

### ISSUE THREAD

**Issue by rviscomi** [HTTPArchive/httparchive.org#368]
@bkardell is interested to track HTML element popularity over time, similar to the analysis he did in the [2019 Markup](https://almanac.httparchive.org/en/2019/markup#perspective-on-value-and-usage) chapter of the Web Almanac. See the [2019 query](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2019/markup/03_02a.sql) for reference. Let's use this issue to discuss what the data needs are and how we could implement an appropriate solution.

Open questions:

1. How often should we analyze popularity changes?
2. Where should the processed data live?
3. How will the results be visualized?

For 1, we rerun the analysis annually for the Web Almanac, so I wonder if we could piggyback off of that to get updated data. Otherwise, if new data is needed as frequently as monthly, we'd need a new pipeline. @bkardell what did you have in mind?

For 2, we'd need a dedicated table in BigQuery, but right now the closest thing we have is the `scratchspace` dataset, which is meant more for temporary storage. Maybe we need an `experimental` dataset for longer-lived projects like this.

For 3, @bkardell I assume you'll reuse [rainy-periwinkle.glitch.me](https://rainy-periwinkle.glitch.me/)? Or should we have a report on [httparchive.org](https://httparchive.org/)?

Any other considerations?

**Comment by tunetheweb** [comment]
For a very quick method, rather than storing this in BigQuery with a dedicated table, we could just generate the data from that query periodically and host a JSON file from our CDN – like we do other queries that run monthly? The JSON could be stored in the monthly folders along with the monthly histograms JSONs.

Then, to visualise, could either point the existing glitch to the httparchive CDN or alternatively could create a Google Data Studio report to read that (similar to what you did recently for CWV Technology report)? Might just need to update our CORS header to allow that access.

Longer term it would be nice to have ability to add more reports like this to httparchive.org rather than just timeseries and histograms but since we don't have anything like that then would need some time investment to set that up...

**Comment by bkardell** [comment]
Yeah, more or less I am just adding the output of those as json to this glitch as files because it's very fast (instant) and cheap to analyze them many ways then. The glitch lets you do fuzzy searches on a dataset/file based on regexps and then gives you a report which you can optionally create a perma-link to.  But it has some other end points like a way to track deltas in them too - https://rainy-periwinkle.glitch.me/delta/mobile/aug_2019/march_2020 and I would love to just be able to continue to hack away on things and develop interesting insights and views, maybe even something with push notifications for me - but every time I get a chance to go back to it though, my info is out of date and I burn some time updating.  It would be very handy to just at least store the json in a somewhat queryable form (meaning "what files exist") so that I can sync it automatically with the most recent data, even if it means subsetting or something.  Currently the files look like https://rainy-periwinkle.glitch.me/aug_2019.json 

Of course, I'm happy to help contribute to larger tooling here too if we're interested but really just getting the data would be great - making it available in csv would be great too as you can import that into so many things already and if we have some flat data like this it's pretty easy to convert.

**Comment by rviscomi** [comment]
> Then, to visualise, could either point the existing glitch to the httparchive CDN or alternatively could create a Google Data Studio report to read that (similar to what you did recently for CWV Technology report)? Might just need to update our CORS header to allow that access.

WDYT about adding a query to the `histograms` directory so that it gets run every month? We don't necessarily need a corresponding metric in the website UI, so the JSON data would still get uploaded to the CDN. 

I ran an ad hoc test for the June dataset: https://cdn.httparchive.org/reports/2021_06_01/htmlElementPopularity.json

@bkardell does that JSON file give you what you need for the time being?

**Comment by tunetheweb** [comment]
Yup seems like the way to go.

Would this work for you @bkardell ? Would you then have a job that copies this to glitch periodically? Or would you read it directly from our CDN (when as I say there might be CORs issues to deal with).

The other advantage of this is that you'd get the lenses. Currently these are the same queries run and then filtering by popular CMS's (Wordpress, Drupal and Magento) but in #377 we're going to add ranking (and potentially deprecate the CMS lenses).

That means you would get separate JSON files at:
-  reports/2021_06_01/htmlElementPopularity.json - using all 6.5 - 7.5 million sites we track
- reports/top1k/2021_06_01/htmlElementPopularity.json - top 1,000 sites only
- reports/top10k/2021_06_01/htmlElementPopularity.json - top 10,000 sites only

So you could see how popularity in top sites differs from the rest of the web if you want to drag in those additional JSON files.

**Comment by bkardell** [comment]
Thanks for the new file @rviscomi !  I've pulled it in - such interesting data and I'd like to make more views and things (more on this below)

To @tunetheweb's q's... Those variants would be interesting, sure.  If I were to try to use directly them  somehow vs import, I'd also need to be able to list what is and isn't there to let people choose, is there a way?  I think that _for now_ I'd like to keep pulling them in manually and experimenting and continuting to think about what we'd like to add to the archive as we get a little more sure... Maybe this tool, or parts of it, or something like it should be part of the archive, but I'm just not sure what yet...

My thing currently only takes the top 10,001 in its dataset (the extra 1 is a mistake in my initial set that I just carried on)... This puts the .json file ~1m or less and keeps things speedy and cheap to process, and results are still somewhat readable. Working with this myself in 2019- the main file is big and way too noisy to really be useful I think.  For any kind of real insight it feels like you gotta break into that as it's a few orders of magnitude beyond how many real elements there are already.  Since these are sorted by pop, I just slice the first 10k (+1 :( ) entries of whatever set @rviscomi has been giving me if it is bigger (the one he provided here is >40mb) - but it also gives me some interesting flexibility as I sort thought... Like, currently the deltas reports like I shared above are also live processed and also based on the subset report, but really the more I think about it the more I should make generation and some analysis part of static processing just once and off the full set.  In any given set, so far, it seems about 1300-1400 "new" elements appear ([1](https://rainy-periwinkle.glitch.me/delta/mobile/aug_2019/march_2020), [2](https://rainy-periwinkle.glitch.me/delta/mobile/march_2020/july_2021)), but they are all ~ the same use, so maybe they're all just kind of middling around a similar spot - that would tell us.

Another useful thing we can do with this data is I have a script that identified "probably errors" and "probably common cause".  While I was writing the 2019 report, I did some investigation on these with rick and we were able to fix a couple thousand markup errors with a single email... It would be nice to make that easy to surface somehow - usually they're escaping errors or something.

Idk - monthy is pretty frequent to manage manually and it might be too much data, but if it is easy to do it's definitely definitely useful to have the info and see what we can imagine doing with it.

**Comment by tunetheweb** [comment]
@rviscomi ’s on vacation so not sure if he’ll see this For a bit.

We have a lot of these queries that run monthly. It’s what drives all the reports at: https://httparchive.org/reports

So very easy for us to add them there. And these I had a look and this query is quick and cheap compared to some we have so no bother adding them. And, as I say, you get the lenses versions for free if you want them.

In fact doing that at any cadence other than monthly is actually **more** work for us as means we couldn’t use the current monthly  process to run the reports automatically.

You can of course import them or not as you see fit, but once we add them to our batch you will always have the data already available (on about the 25th of the month) for when you want to refresh it.

I can also easily limit the number of rows returned if you only want the top 10,001 rows to make the files smaller? Or would you prefer to leave it at the full dataset in case you want it and cut it down yourself?

Anyway let me know if you’re good for me to add the query to the batch and will CC you on the PR.

**Comment by rviscomi** [comment]
@tunetheweb this was the file I used for testing: `/sql/histograms/htmlElementPopularity.sql`

```sql
#standardSQL
CREATE TEMPORARY FUNCTION getElements(payload STRING)
RETURNS ARRAY<STRING> LANGUAGE js AS '''
try {
  var $ = JSON.parse(payload);
  var elements = JSON.parse($._element_count);
  if (Array.isArray(elements) || typeof elements != 'object') return [];
  return Object.keys(elements);
} catch (e) {
  return [];
}
''';

 SELECT
  _TABLE_SUFFIX AS client,
  element,
  COUNT(DISTINCT url) AS pages,
  total,
  COUNT(DISTINCT url) / total AS pct
FROM
  `httparchive.pages.${YYYY_MM_DD}_*`
JOIN
  (SELECT _TABLE_SUFFIX, COUNT(0) AS total FROM `httparchive.pages.${YYYY_MM_DD}_*` GROUP BY _TABLE_SUFFIX)
USING (_TABLE_SUFFIX),
  UNNEST(getElements(payload)) AS element
GROUP BY
  client,
  total,
  element
ORDER BY
  pages / total DESC,
  client
```

**Comment by bkardell** [comment]
@tunetheweb yeah I'm ok with any version of that (i can cut it down myself, or having a cut down one would save some space and there's diminishing value.. another way to do that would be to filter results with more than some small minimum.. i expect the full set includes a bunch of elements that occur on 1 or 2 sites.. not especially useful to know.. maybe more than 10 or something).

i know that in the past, one thing I've wanted to know is a sample of a few urls that had these.. not sure if we can make that easier somehow too but it would be great too

**Comment by tunetheweb** [comment]
OK added those both and new test file available at: https://storage.googleapis.com/httparchive/reports/2021_07_01/htmlElementPopularity.json

@rviscomi 's original run was 45MB. Limiting it to elements on 10 or more sites brought that down considerably, but then adding in the 5 sample URLs brought it back up. It's now 6.7 MB - so still a lot smaller than in the past.

It includes this one that you were asking about on Slack:

```json
{
"client": "mobile",
"element": "include",
"pages": "5600",
"pct": "7.474799248248761E-4",
"sample_urls": "https://indonesiwa.blogspot.com/ https://www.zawaml.tk/ https://www.apkamigo.com/ http://www.infoinhu.com/ https://www.usamatech7.com/",
"total": "7491840"
},
```

**Comment by tunetheweb** [comment]
Any thoughts @bkardell ? Am I good to merge that in now?
